File: proton-c/bindings/ruby/lib/util/wrapper.rb
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.


module Qpid::Proton
  module Util

    # @private
    # Class methods to help wrapper classes define forwarding methods to proton-C functions
    #
    # The extending class must define PROTON_METHOD_PREFIX, the functions here
    # make it easy to define ruby methods to forward calls to C functions.
    #
    module SWIGClassHelper
      # Define ruby method +name+ to forward arguments to
      # CProton.PROTON_METHOD_PREFIX_+pn_name+(@impl, ...)
      def proton_forward(name, pn_name)
        pn_name = pn_name[0..-2] if pn_name.to_s.end_with? "?" # Drop trailing ? for ruby bool methods
        pn_name = "#{self::PROTON_METHOD_PREFIX}_#{pn_name}".to_sym
        define_method(name.to_sym) { |*args| Cproton.__send__(pn_name, @impl, *args) }
      end

      def proton_caller(*names) names.each { |name| proton_forward(name, name) }; end
      def proton_set(*names) names.each { |name| proton_forward("#{name}=", "set_#{name}") }; end
      def proton_get(*names) names.each { |name| proton_forward(name, "get_#{name}") }; end
      def proton_is(*names) names.each { |name| proton_forward("#{name}?", "is_#{name}") }; end
      def proton_set_get(*names) names.each { |name| proton_get(name); proton_set(name) }; end
      def proton_set_is(*names) names.each { |name| proton_is(name); proton_set(name) }; end

      # Store ruby wrappers as attachments so they can be retrieved from the C pointer.
      #
      # Wrappers are stored in a registry using a key. The key is then attached to
      # the Proton structure as a record. That record lives for as long as the
      # Proton structure lives, and when the structure is released the record acts
      # as hook to also delete the Ruby wrapper object from the registry.

      @@registry = {}

      # @private
      def get_key(impl)
        ("%032x" % Cproton.pni_address_of(impl))
      end

      # @private
      # Stores the given object for later retrieval.
      #
      # @param object [Object] The object.
      # @param attachment_method [Symbol] The Proton attachment method.
      #
      def store_instance(object, attachment_method = nil)
        # ensure the impl has a reference to the wrapper object
        object.impl.instance_eval { @proton_wrapper = object }
        registry_key = get_key(object.impl)
        unless attachment_method.nil?
          record = Cproton.__send__(attachment_method, object.impl)
          rbkey = Cproton.Pn_rbkey_new
          Cproton.Pn_rbkey_set_registry(rbkey, Cproton.pn_rb2void(Qpid::Proton::Util::Wrapper.registry))
          Cproton.Pn_rbkey_set_method(rbkey, "delete")
          Cproton.Pn_rbkey_set_key_value(rbkey, registry_key)
          Cproton.pn_record_def(record, RBCTX, Cproton.Pn_rbkey__class());
          Cproton.pn_record_set(record, RBCTX, rbkey)
        end
        @@registry[registry_key] = object
      end

      # Retrieves the wrapper object with the supplied Proton struct.
      #
      # @param impl [Object] The wrapper for the Proton struct.
      # @param attachment_method [Symbol] The Proton attachment method.
      #
      # @return [Object] The Ruby wrapper object.
      #
      def fetch_instance(impl, attachment_method = nil)
        # if the impl has a wrapper already attached, then return it
        if impl.instance_variable_defined?(:@proton_wrapper)
          return impl.instance_variable_get(:@proton_wrapper)
        end
        unless attachment_method.nil?
          record = Cproton.__send__(attachment_method, impl)
          rbkey = Cproton.pni_void2rbkey(Cproton.pn_record_get(record, RBCTX))
          # if we don't have a key, then we don't have an object
          return nil if rbkey.nil?
          registry_key = Cproton.Pn_rbkey_get_key_value(rbkey)
        else
          registry_key = get_key(impl)
        end
        # if the object's not in the registry then return
        return nil unless @@registry.has_key?(registry_key)

        result = @@registry[registry_key]
        # result = nil unless result.weakref_alive?
        if result.nil?
          raise Qpid::Proton::ProtonError.new("missing object for key=#{registry_key}")
        else
          # update the impl since the Swig wrapper for it may have changed
          result.impl = impl
        end
        return result
      end
      RBCTX = self.hash.to_i
    end

    # @private
    #
    # Instance methods to include in classes that wrap pn_object types
    # that support pn_inspect etc. Automatically extends SWIGClassHelper
    #
    module Wrapper

      def self.included(base)
        base.extend(SWIGClassHelper)
      end

      attr_accessor :impl

      def inspect
        return "#{self.class}<nil>" unless @impl
        pstr = Cproton.pn_string("")
        begin
          Cproton.pn_inspect(@impl, pstr)
          return Cproton.pn_string_get(pstr)
        ensure
          Cproton.pn_free(pstr)
        end
      end

      def to_s() inspect; end

      def self.registry
        @registry ||= {}
      end
    end
  end
end

File: python/proton/_wrapper.py
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#

from typing import Any, Callable, ClassVar, Optional

from cproton import addressof, isnull, pn_incref, pn_decref, \
    pn_record_get_py

from ._exceptions import ProtonException


class Wrapper:
    """ Wrapper for python objects that need to be stored in event contexts and be retrieved again from them
        Quick note on how this works:
        The actual *python* object has only 2 attributes which redirect into the wrapped C objects:
        _impl   The wrapped C object itself
        _attrs  This is a special pn_record_t holding a PYCTX which is a python dict
                every attribute in the python object is actually looked up here

        Because the objects actual attributes are stored away they must be initialised *after* the wrapping. This is
        achieved by using the __new__ method of Wrapper to create the object with the actiual attributes before the
        __init__ method is called.

        In the case where an existing object is being wrapped, the __init__ method is called with the existing object
        but should not initialise the object. This is because the object is already initialised and the attributes
        are already set. Use the Uninitialised method to check if the object is already initialised.
    """

    constructor: ClassVar[Optional[Callable[[], Any]]] = None
    get_context: ClassVar[Optional[Callable[[Any], dict[str, Any]]]] = None

    __slots__ = ["_impl", "_attrs"]

    @classmethod
    def wrap(cls, impl: Any) -> Optional['Wrapper']:
        if isnull(impl):
            return None
        else:
            return cls(impl)

    def __new__(cls, impl: Any = None) -> 'Wrapper':
        attrs = None
        try:
            if impl is None:
                # we are constructing a new object
                assert cls.constructor
                impl = cls.constructor()
                if impl is None:
                    raise ProtonException(
                        "Wrapper failed to create wrapped object. Check for file descriptor or memory exhaustion.")
            else:
                # we are wrapping an existing object
                pn_incref(impl)

            assert cls.get_context
            record = cls.get_context(impl)
            attrs = pn_record_get_py(record)
        finally:
            self = super().__new__(cls)
            self._impl = impl
            self._attrs = attrs
            return self

    def Uninitialized(self) -> bool:
        return self._attrs == {}

    def __getattr__(self, name: str) -> Any:
        attrs = self._attrs
        if attrs and name in attrs:
            return attrs[name]
        else:
            raise AttributeError(name + " not in _attrs")

    def __setattr__(self, name: str, value: Any) -> None:
        if hasattr(self.__class__, name):
            object.__setattr__(self, name, value)
        else:
            attrs = self._attrs
            if attrs is not None:
                attrs[name] = value

    def __delattr__(self, name: str) -> None:
        attrs = self._attrs
        if attrs:
            del attrs[name]

    def __hash__(self) -> int:
        return hash(addressof(self._impl))

    def __eq__(self, other: Any) -> bool:
        if isinstance(other, Wrapper):
            return addressof(self._impl) == addressof(other._impl)
        return False

    def __ne__(self, other: Any) -> bool:
        if isinstance(other, Wrapper):
            return addressof(self._impl) != addressof(other._impl)
        return True

    def __del__(self) -> None:
        pn_decref(self._impl)

    def __repr__(self) -> str:
        return '<%s.%s 0x%x ~ 0x%x>' % (self.__class__.__module__,
                                        self.__class__.__name__,
                                        id(self), addressof(self._impl))
